src/playlist-loader/dash-main-playlist-loader.js

```diff
@@ -0,0 +1,179 @@
+import PlaylistLoader from './playlist-loader.js';
+import {resolveUrl} from '../resolve-url';
+import {parse as parseMpd, parseUTCTiming} from 'mpd-parser';
+import {mergeManifest, forEachPlaylist} from './utils.js';
+
+/**
+ * An instance of the `DashMainPlaylistLoader` class is created when VHS is passed a DASH
+ * manifest. For dash main playlists are the only thing that needs to be refreshed. This
+ * is important to note as a lot of the `DashMediaPlaylistLoader` logic looks to
+ * `DashMainPlaylistLoader` for guidance.
+ *
+ * @extends PlaylistLoader
+ */
+class DashMainPlaylistLoader extends PlaylistLoader {
+
+  /**
+   * Create an instance of this class.
+   *
+   * @param {Element} uri
+   *        The uri of the manifest.
+   *
+   * @param {Object} options
+   *        Options that can be used, see the base class for more information.
+   */
+  constructor(uri, options) {
+    super(uri, options);
+    this.clientOffset_ = null;
+    this.clientClockOffset_ = null;
+    this.setMediaRefreshTimeout_ = this.setMediaRefreshTimeout_.bind(this);
+  }
+
+  /**
+   * Get an array of all playlists in this manifest, including media group
+   * playlists.
+   *
+   * @return {Object[]}
+   *          An array of playlists.
+   */
+  playlists() {
+    const playlists = [];
+
+    forEachPlaylist(this.manifest_, (media) => {
+      playlists.push(media);
+    });
+
+    return playlists;
+  }
+
+  /**
+   * Parse a new manifest and merge it with an old one. Calls back
+   * with the merged manifest and weather or not it was updated.
+   *
+   * @param {string} manifestString
+   *        A manifest string directly from the request response.
+   *
+   * @param {Function} callback
+   *        A callback that takes the manifest and updated
+   *
+   * @private
+   */
+  parseManifest_(manifestString, callback) {
+    this.syncClientServerClock_(manifestString, (clientOffset) => {
+      const parsedManifest = parseMpd(manifestString, {
+        manifestUri: this.uri_,
+        clientOffset
+      });
+
+      // merge everything except for playlists, they will merge themselves
+      const mergeResult = mergeManifest(this.manifest_, parsedManifest, ['playlists', 'mediaGroups']);
```

Comment:
Not also that we do not merge the `playlists` or `mediaGroups` here, which means only the new unmerged playlists will exist after this point. We do this because each individual DashMediaPlaylistLoader merges it's own playlist (if the playlist loader is started).